File: scripts/test_stock_in_gl.py
#!/usr/bin/env python3
"""
Test script for Stock-In functionality with GL integration
Tests the complete flow from stock movement to journal entry creation
"""

import requests
import json
import time
from datetime import datetime
from typing import Dict, Any, List

# Configuration
API_BASE = "http://localhost:3001/api"
HEADERS = {"Content-Type": "application/json"}

# Test items
TEST_ITEMS = [
    {
        "itemCode": "LAPTOP-001",
        "quantity": 5,
        "unitCost": 1500.00,
        "location": "WAREHOUSE-A",
        "reference": f"TEST-LAPTOP-{int(time.time())}",
        "notes": "Test laptop stock-in with GL integration"
    },
    {
        "itemCode": "PAPER-A4",
        "quantity": 100,
        "unitCost": 5.00,
        "location": "WAREHOUSE-A",
        "reference": f"TEST-PAPER-{int(time.time())}",
        "notes": "Test paper stock-in with GL integration"
    }
]


def api_call(endpoint: str, method: str = "GET", data: Dict[str, Any] = None) -> Dict[str, Any]:
    """Make API call and return response data"""
    url = f"{API_BASE}{endpoint}"
    
    try:
        if method == "GET":
            response = requests.get(url, headers=HEADERS)
        elif method == "POST":
            response = requests.post(url, headers=HEADERS, json=data)
        else:
            raise ValueError(f"Unsupported method: {method}")
        
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"❌ API Error: {e}")
        if hasattr(e.response, 'text'):
            print(f"Response: {e.response.text}")
        raise


def test_stock_in(item_config: Dict[str, Any]) -> Dict[str, Any]:
    """Test a single stock-in transaction"""
    print(f"\n🧪 Testing Stock-In for {item_config['itemCode']}")
    print("─" * 50)
    
    result = {"success": False, "item": item_config['itemCode']}
    
    try:
        # Step 1: Check initial balance
        print("📊 Checking initial inventory balance...")
        balance_response = api_call(
            f"/inventory/balances?itemCode={item_config['itemCode']}&location={item_config['location']}"
        )
        initial_balance = balance_response.get("data", [{}])[0].get("quantity", 0)
        print(f"  Initial balance: {initial_balance}")
        
        # Step 2: Create stock-in movement
        print("\n📥 Creating stock-in transaction...")
        movement_data = {
            "type": "IN",
            "itemCode": item_config["itemCode"],
            "quantity": item_config["quantity"],
            "unitCost": item_config["unitCost"],
            "location": item_config["location"],
            "reference": item_config["reference"],
            "notes": item_config["notes"]
        }
        
        movement_response = api_call("/stock-movements", "POST", movement_data)
        movement_id = movement_response["data"]["id"]
        total_value = item_config["quantity"] * item_config["unitCost"]
        
        print(f"  ✅ Movement created: ID {movement_id}")
        print(f"  Total value: ${total_value:,.2f}")
        result["movement_id"] = movement_id
        
        # Step 3: Verify inventory update
        print("\n📈 Verifying inventory balance update...")
        time.sleep(1)  # Wait for processing
        
        new_balance_response = api_call(
            f"/inventory/balances?itemCode={item_config['itemCode']}&location={item_config['location']}"
        )
        new_balance = new_balance_response.get("data", [{}])[0].get("quantity", 0)
        expected_balance = initial_balance + item_config["quantity"]
        
        print(f"  New balance: {new_balance}")
        print(f"  Expected: {expected_balance}")
        
        if new_balance == expected_balance:
            print("  ✅ Balance updated correctly")
        else:
            print("  ❌ Balance mismatch!")
            
        # Step 4: Check journal entry
        print("\n📚 Checking journal entry creation...")
        journal_response = api_call(f"/journal-entries?reference=STOCK-IN-{movement_id}")
        
        if journal_response.get("data"):
            entry = journal_response["data"][0]
            print(f"  ✅ Journal entry found: ID {entry['id']}")
            print(f"  Date: {entry['date']}")
            print(f"  Description: {entry['description']}")
            
            # Analyze journal lines
            print("\n  Journal Lines:")
            total_debits = 0
            total_credits = 0
            
            for line in entry["lines"]:
                account_name = line.get("account", {}).get("name", line["accountCode"])
                amount = float(line["amount"])
                print(f"    {line['type']}: {account_name} - ${amount:,.2f}")
                
                if line["type"] == "DEBIT":
                    total_debits += amount
                else:
                    total_credits += amount
                
                # Verify amount matches expected
                if abs(amount - total_value) > 0.01:
                    print(f"    ⚠️  Amount mismatch! Expected: ${total_value:,.2f}")
            
            # Verify balanced entry
            print(f"\n  Total Debits: ${total_debits:,.2f}")
            print(f"  Total Credits: ${total_credits:,.2f}")
            
            if abs(total_debits - total_credits) < 0.01:
                print("  ✅ Entry is balanced")
                result["success"] = True
            else:
                print("  ❌ Entry is not balanced!")
        else:
            print("  ❌ No journal entry found!")
            
    except Exception as e:
        print(f"  ❌ Error: {e}")
        result["error"] = str(e)
    
    return result


def test_gl_balances():
    """Test GL account balances"""
    print("\n💰 Checking GL Account Balances")
    print("─" * 50)
    
    try:
        # Check inventory account (1300)
        print("Inventory Account (1300):")
        inv_response = api_call("/accounts/1300/balance")
        print(f"  Balance: ${inv_response.get('balance', 0):,.2f}")
        
        # Check accounts payable (2100)
        print("\nAccounts Payable (2100):")
        ap_response = api_call("/accounts/2100/balance")
        print(f"  Balance: ${ap_response.get('balance', 0):,.2f}")
        
    except Exception as e:
        print(f"  Error checking balances: {e}")


def run_tests():
    """Main test runner"""
    print("🚀 Starting Stock-In GL Integration Tests")
    print("═" * 50)
    print(f"Server: {API_BASE}")
    print(f"Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    
    results = []
    
    # Test each item
    for item_config in TEST_ITEMS:
        result = test_stock_in(item_config)
        results.append(result)
    
    # Check GL balances
    test_gl_balances()
    
    # Summary
    print("\n📊 Test Summary")
    print("═" * 50)
    
    successful = sum(1 for r in results if r.get("success"))
    failed = len(results) - successful
    
    print(f"Total tests: {len(results)}")
    print(f"✅ Successful: {successful}")
    print(f"❌ Failed: {failed}")
    
    if failed > 0:
        print("\nFailed tests:")
        for r in results:
            if not r.get("success"):
                print(f"  - {r['item']}: {r.get('error', 'Unknown error')}")
    
    print("\n✨ Test run complete!")


if __name__ == "__main__":
    run_tests()

File: scripts/test-stock-in-gl.js
#!/usr/bin/env node

/**
 * Test script for Stock-In functionality with GL integration
 * This script tests the complete flow from stock movement to journal entry creation
 */

const API_BASE = 'http://localhost:3000/api';

// Test configuration
const TEST_CONFIG = {
  items: [
    {
      itemCode: 'LAPTOP-001',
      quantity: 5,
      unitCost: 1500.00,
      location: 'WAREHOUSE-A',
      reference: `TEST-LAPTOP-${Date.now()}`,
      notes: 'Test laptop stock-in with GL'
    },
    {
      itemCode: 'PAPER-A4',
      quantity: 100,
      unitCost: 5.00,
      location: 'WAREHOUSE-A',
      reference: `TEST-PAPER-${Date.now()}`,
      notes: 'Test paper stock-in with GL'
    }
  ]
};

// Helper function to make API calls
async function apiCall(endpoint, options = {}) {
  const url = `${API_BASE}${endpoint}`;
  const response = await fetch(url, {
    headers: {
      'Content-Type': 'application/json',
      ...options.headers
    },
    ...options
  });
  
  const data = await response.json();
  if (!response.ok) {
    throw new Error(`API Error: ${data.error || response.statusText}`);
  }
  return data;
}

// Test a single stock-in transaction
async function testStockIn(config) {
  console.log(`\n🧪 Testing Stock-In for ${config.itemCode}`);
  console.log('─'.repeat(50));
  
  try {
    // Step 1: Check initial balance
    console.log('📊 Checking initial inventory balance...');
    const initialBalance = await apiCall(`/inventory/balances?itemCode=${config.itemCode}&location=${config.location}`);
    const currentQty = initialBalance.data?.[0]?.quantity || 0;
    console.log(`  Initial balance: ${currentQty}`);
    
    // Step 2: Create stock-in movement
    console.log('\n📥 Creating stock-in transaction...');
    const stockInData = {
      type: 'IN',
      itemCode: config.itemCode,
      quantity: config.quantity,
      unitCost: config.unitCost,
      location: config.location,
      reference: config.reference,
      notes: config.notes
    };
    
    const movement = await apiCall('/stock-movements', {
      method: 'POST',
      body: JSON.stringify(stockInData)
    });
    
    console.log(`  ✅ Movement created: ID ${movement.data.id}`);
    console.log(`  Total value: $${(config.quantity * config.unitCost).toFixed(2)}`);
    
    // Step 3: Verify inventory update
    console.log('\n📈 Verifying inventory balance update...');
    await new Promise(resolve => setTimeout(resolve, 1000)); // Wait for processing
    
    const newBalance = await apiCall(`/inventory/balances?itemCode=${config.itemCode}&location=${config.location}`);
    const newQty = newBalance.data?.[0]?.quantity || 0;
    const expectedQty = currentQty + config.quantity;
    
    console.log(`  New balance: ${newQty}`);
    console.log(`  Expected: ${expectedQty}`);
    console.log(`  ${newQty === expectedQty ? '✅ Balance updated correctly' : '❌ Balance mismatch!'}`);
    
    // Step 4: Check journal entry
    console.log('\n📚 Checking journal entry creation...');
    const journalEntries = await apiCall(`/journal-entries?reference=STOCK-IN-${movement.data.id}`);
    
    if (journalEntries.data && journalEntries.data.length > 0) {
      const entry = journalEntries.data[0];
      console.log(`  ✅ Journal entry found: ID ${entry.id}`);
      console.log(`  Date: ${new Date(entry.date).toLocaleDateString()}`);
      console.log(`  Description: ${entry.description}`);
      
      // Analyze journal lines
      console.log('\n  Journal Lines:');
      const totalAmount = config.quantity * config.unitCost;
      
      entry.lines.forEach(line => {
        const accountName = line.account?.name || line.accountCode;
        console.log(`    ${line.type}: ${accountName} - $${line.amount}`);
        
        // Verify amounts
        if (Math.abs(parseFloat(line.amount) - totalAmount) > 0.01) {
          console.log(`    ⚠️  Amount mismatch! Expected: $${totalAmount}`);
        }
      });
      
      // Verify balanced entry
      const totalDebits = entry.lines
        .filter(l => l.type === 'DEBIT')
        .reduce((sum, l) => sum + parseFloat(l.amount), 0);
      const totalCredits = entry.lines
        .filter(l => l.type === 'CREDIT')
        .reduce((sum, l) => sum + parseFloat(l.amount), 0);
      
      console.log(`\n  Total Debits: $${totalDebits.toFixed(2)}`);
      console.log(`  Total Credits: $${totalCredits.toFixed(2)}`);
      console.log(`  ${Math.abs(totalDebits - totalCredits) < 0.01 ? '✅ Entry is balanced' : '❌ Entry is not balanced!'}`);
      
    } else {
      console.log('  ❌ No journal entry found!');
    }
    
    return { success: true, movementId: movement.data.id };
    
  } catch (error) {
    console.error(`  ❌ Error: ${error.message}`);
    return { success: false, error: error.message };
  }
}

// Test GL account balances
async function testGLBalances() {
  console.log('\n💰 Checking GL Account Balances');
  console.log('─'.repeat(50));
  
  try {
    // Check inventory account (1300)
    console.log('Inventory Account (1300):');
    const inventoryBalance = await apiCall('/accounts/1300/balance');
    console.log(`  Balance: $${inventoryBalance.balance || 0}`);
    
    // Check accounts payable (2100)
    console.log('\nAccounts Payable (2100):');
    const payableBalance = await apiCall('/accounts/2100/balance');
    console.log(`  Balance: $${payableBalance.balance || 0}`);
    
  } catch (error) {
    console.error(`  Error checking balances: ${error.message}`);
  }
}

// Main test runner
async function runTests() {
  console.log('🚀 Starting Stock-In GL Integration Tests');
  console.log('═'.repeat(50));
  console.log(`Server: ${API_BASE}`);
  console.log(`Time: ${new Date().toLocaleString()}`);
  
  const results = [];
  
  // Test each item
  for (const config of TEST_CONFIG.items) {
    const result = await testStockIn(config);
    results.push({ ...config, ...result });
  }
  
  // Check GL balances
  await testGLBalances();
  
  // Summary
  console.log('\n📊 Test Summary');
  console.log('═'.repeat(50));
  
  const successful = results.filter(r => r.success).length;
  const failed = results.filter(r => !r.success).length;
  
  console.log(`Total tests: ${results.length}`);
  console.log(`✅ Successful: ${successful}`);
  console.log(`❌ Failed: ${failed}`);
  
  if (failed > 0) {
    console.log('\nFailed tests:');
    results.filter(r => !r.success).forEach(r => {
      console.log(`  - ${r.itemCode}: ${r.error}`);
    });
  }
  
  console.log('\n✨ Test run complete!');
}

// Run the tests
runTests().catch(console.error);